contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,380 @@
+
+* CreateSnapshot, DeleteSnapshot, and create volume from snapshot if it supports CREATE_DELETE_SNAPSHOT.
+* ListSnapshots if it supports LIST_SNAPSHOTS.
+
+ListSnapshots can be an expensive operation because it will try to list all snapshots on the storage system. For a storage system that takes nightly periodic snapshots, the total number of snapshots on the system can be huge. Kubernetes should try to avoid this call if possible. Instead, calling ListSnapshots with a specific snapshot_id as filtering to query the status of the snapshot will be more desirable and efficient.
+
+CreateSnapshot is a synchronous function and it must be blocking until the snapshot is cut. For cloud providers that support the uploading of a snapshot as part of the creating snapshot operation, CreateSnapshot function must also be blocking until the snapshot is cut and after that it shall return an operation pending gRPC error code until the uploading process is complete.
+
+Create volume from snapshot will be handled by the CreateVolume controller function in the CSI Volume Driver.
+
+Refer to [Container Storage Interface (CSI)](https://github.com/container-storage-interface/spec) for detailed instructions on how CSI Volume Driver shall implement snapshot functions.
+
+
+## Transition to the New Snapshot Support
+
+To use the snapshot implementation in [external storage repo](https://github.com/kubernetes-incubator/external-storage/tree/master/snapshot), an external snapshot controller and external provisioner need to bedeployed.
+
+To use the new snapshot implementation for CSI, a sidecar container for the external snapshot controller needs to be deployed in addition to the sidecar container for the external provisioner.
+
+The differences between the two are as follows from a user's point of view:
+
+New Design
+* Supports CSI volume drivers.
+* To create a snapshot for CSI, a SnapshotClass can be created and specified in the spec of VolumeSnapshot.
+* To restore a volume from the snapshot, user should use the same StorageClass that is used for the original PVC.
+
+Old Design
+* Does not support CSI volume drivers.
+* SnapshotClass is not needed to create a snapshot and this concept does not exist in the old design.
+* To restore a volume from the snapshot, however, user needs to create a new StorageClass that is different from the original one for the PVC.
+
+Here is an example yaml file to create a snapshot in the old design:
+
+```GO
+
+apiVersion: volumesnapshot.external-storage.k8s.io/v1
+kind: VolumeSnapshot
+metadata:
+  name: hostpath-test-snapshot
+spec:
+  persistentVolumeClaimName: pvc-test-hostpath
+
+```
+
+Here is an example to create a SnapshotClass and create a snapshot in the new design:
+
+```GO
+
+apiVersion: volumesnapshot.csi.k8s.io/v1alpha1
+kind: SnapshotClass
+metadata:
+  name: csi-hostpath-snapclass
+  snapshotter: csi-hostpath
```

Comment:
the format maybe wrong, remove the white space

Reply:
Ok, I'll check.